File: api/src/main/java/fr/huiitre/tools/modules/riot/valorant/infrastructure/PostgresValorantBundleRepository.java
package fr.huiitre.tools.modules.riot.valorant.infrastructure;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;

import fr.huiitre.tools.modules.riot.valorant.application.catalog.ports.ValorantBundleRepository;
import fr.huiitre.tools.modules.riot.valorant.application.catalog.view.ValorantBundleView;

public class PostgresValorantBundleRepository implements ValorantBundleRepository {

    private final JdbcTemplate jdbcTemplate;

    public PostgresValorantBundleRepository(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

    private static final RowMapper<ValorantBundleView> ROW_MAPPER = (rs, rowNum) -> new ValorantBundleView(
            rs.getLong("id"),
            rs.getObject("asset_id", UUID.class),
            rs.getString("name"),
            rs.getString("banner_url"));

    @Override
    public List<ValorantBundleView> findAll() {
        final String sql = """
                SELECT id, asset_id, name, banner_url
                FROM tools_riot.valorant_bundles
                ORDER BY name
                """;
        return jdbcTemplate.query(sql, ROW_MAPPER);
    }

    @Override
    public Optional<ValorantBundleView> findById(Long id) {
        final String sql = """
                SELECT id, asset_id, name, banner_url
                FROM tools_riot.valorant_bundles
                WHERE id = ?
                """;
        return jdbcTemplate.query(sql, ROW_MAPPER, id).stream().findFirst();
    }

    @Override
    public Optional<ValorantBundleView> findByAssetId(UUID assetId) {
        final String sql = """
                SELECT id, asset_id, name, banner_url
                FROM tools_riot.valorant_bundles
                WHERE asset_id = ?
                """;
        return jdbcTemplate.query(sql, ROW_MAPPER, assetId).stream().findFirst();
    }
}

File: api/Modules/Riot/Valorant/Infrastructure/PostgresValorantBundleRepository.cs
using Tools.Api.Modules.Riot.Common.Infrastructure;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Ports;
using Tools.Api.Modules.Riot.Valorant.Application.Catalog.Views;

namespace Tools.Api.Modules.Riot.Valorant.Infrastructure;

public sealed class PostgresValorantBundleRepository(RiotDatabase database) : IValorantBundleRepository
{
    private const string Select = """
        SELECT id AS Id, asset_id AS AssetId, name AS Name, banner_url AS BannerUrl
        FROM tools_riot.valorant_bundles
        """;

    public Task<List<ValorantBundleView>> FindAll() =>
        database.Query<ValorantBundleView>($"{Select} ORDER BY name");

    public Task<ValorantBundleView?> FindById(long id) =>
        database.QueryFirstOrDefault<ValorantBundleView>($"{Select} WHERE id = @Id", new { Id = id });

    public Task<ValorantBundleView?> FindByAssetId(Guid assetId) =>
        database.QueryFirstOrDefault<ValorantBundleView>($"{Select} WHERE asset_id = @AssetId", new { AssetId = assetId });
}
